prov/efa/src/efa_cq.c

```diff
@@ -121,48 +131,42 @@ static void efa_cq_read_data_entry(struct efa_wc *wc, int i, void *buf)
 ssize_t efa_cq_readfrom(struct fid_cq *cq_fid, void *buf, size_t count,
 			fi_addr_t *src_addr)
 {
+	bool should_end_poll = false;
 	struct efa_cq *cq;
-	struct efa_wce *wce;
-	struct slist_entry *entry;
 	struct efa_av *av;
 	struct efa_wc wc = {0};
-	ssize_t ret = 0, i;
+	ssize_t ret = 0;
+	size_t i = 0;
+
+	/* Initialize an empty ibv_poll_cq_attr struct for ibv_start_poll.
+	 * EFA expects .comp_mask = 0, or otherwise returns EINVAL.
+	 */
+	struct ibv_poll_cq_attr poll_cq_attr = {.comp_mask = 0};
 
 	cq = container_of(cq_fid, struct efa_cq, util_cq.cq_fid);
 
 	ofi_spin_lock(&cq->lock);
 
-	for (i = 0; i < count; i++) {
-		if (!slist_empty(&cq->wcq)) {
-			wce = container_of(cq->wcq.head, struct efa_wce, entry);
-			if (wce->wc.ibv_wc.status) {
-				ret = -FI_EAVAIL;
-				break;
-			}
-			entry = slist_remove_head(&cq->wcq);
-			wce = container_of(entry, struct efa_wce, entry);
-			cq->read_entry(&wce->wc, i, buf);
-			ofi_buf_free(wce);
-			continue;
-		}
-
-		ret = ibv_poll_cq(cq->ibv_cq, 1, &wc.ibv_wc);
-		if (ret != 1) {
-			if (!ret)
+	/* Call ibv_start_poll only once */
+	ret = ibv_start_poll(cq->ibv_cq_ex, &poll_cq_attr);
+	if (!ret)
+		should_end_poll = true;
+
+	do {
```

Comment:
How about `while (!ret && i < cout)`

Reply:
I think `do {xxx} while` is better in this case, because the `{xxx}` block also manipulates the return code from both `ibv_start/next_poll`. We probably do not want to exit the loop if `ret != 0`.

Reply:
I think the part that manipulate return code from `ibv_start/next_poll` can be encapsulated in to an inline function. That function can be used also by `rdm_ep_poll_ibv_cq`.